iree/turbine/kernel/wave/runtime/runtime.cpp

```diff
@@ -1,11 +1,75 @@
+#include <cstring>
 #include <nanobind/nanobind.h>
 #include <nanobind/stl/bind_vector.h>
 #include <nanobind/stl/string.h>
-#include <torch/torch.h>
-#include <torch/extension.h>
-#include <hip/hip_runtime.h>
-#include <hip/hip_runtime_api.h>
-#include <ATen/hip/HIPContext.h>
+#include <sstream>
+
+#ifdef __linux__
+#include <dlfcn.h> // dlopen
+#endif
+
+// Just hardcode necessary constants and types here, we don't expect them to
+// change as it will break all hip programs in existence.
+#define HIP_LAUNCH_PARAM_BUFFER_POINTER ((void*)0x01)
+#define HIP_LAUNCH_PARAM_BUFFER_SIZE ((void*)0x02)
+#define HIP_LAUNCH_PARAM_END ((void*)0x03)
+
+using hipError_t = int;
+using hipStream_t = void*;
+using hipFunction_t = void*;
+using hipModule_t = void*;
+
+ using hipModuleLaunchKernel_t = hipError_t(*)(hipFunction_t,
+    unsigned int, unsigned int,
+    unsigned int, unsigned int,
+    unsigned int, unsigned int,
+    unsigned int, hipStream_t,
+    void**, void**);
+
+using hipGetErrorName_t = const char*(*)(hipError_t);
+using hipGetErrorString_t = const char*(*)(hipError_t);
+using hipModuleUnload_t = hipError_t(*)(hipModule_t);
+using hipModuleLoad_t = hipError_t(*)(hipModule_t*, const char*);
+using hipModuleGetFunction_t = hipError_t(*)(hipFunction_t*, hipModule_t, const char*);
+
+static hipModuleLaunchKernel_t hipModuleLaunchKernel = nullptr;
+static hipGetErrorName_t hipGetErrorName = nullptr;
+static hipGetErrorString_t hipGetErrorString = nullptr;
+static hipModuleUnload_t hipModuleUnload = nullptr;
+static hipModuleLoad_t hipModuleLoad = nullptr;
+static hipModuleGetFunction_t hipModuleGetFunction = nullptr;
+
+static void load_hip_functions()
+{
+    if (hipModuleLaunchKernel && hipGetErrorName && hipGetErrorString &&
+        hipModuleUnload && hipModuleLoad && hipModuleGetFunction)
+        return;
+
+#ifdef __linux__
+#define GET_FUNC(module, name) do { \
+    name = reinterpret_cast<decltype(name)>(dlsym(module, #name)); \
+    if (!name) { \
+        throw std::runtime_error("Failed to load symbol: " + std::string(#name)); \
+    } \
+} while (0)
+    // We expect this module to be loaded permanently in the process so we don't
+    // care about unloading it.
+    auto module = dlopen("libamdhip64.so", RTLD_NOW);
+    if (!module)
+        throw std::runtime_error("Failed to load libamdhip64.so: " + std::string(dlerror()));
+
+    GET_FUNC(module, hipModuleLaunchKernel);
+    GET_FUNC(module, hipGetErrorName);
+    GET_FUNC(module, hipGetErrorString);
+    GET_FUNC(module, hipModuleUnload);
+    GET_FUNC(module, hipModuleLoad);
+    GET_FUNC(module, hipModuleGetFunction);
+
+#undef GET_FUNC
+#else
+    #error "Unsupported platform"
+#endif
```

Comment:
Windows support wouldn't be too difficult to add here. See how https://github.com/iree-org/iree/blob/main/runtime/src/iree/hal/drivers/hip/dynamic_symbols.c searches for `amdhip64.dll` and uses this cross-platform dynamic library loader: https://github.com/iree-org/iree/blob/main/runtime/src/iree/base/internal/dynamic_library.h

Fine to start with only Linux matching the existing support, given how much of an improvement the rest of this PR is.

Reply:
Windows support here is trivial, yeah, the only thing which is stopping me is the we have no way to test it currently.